### DISCUSSION TITLE
PathIcon dont showing inside of Button or ListBoxItem

### DISCUSSION THREAD

**Discussion by alex6dj** [amwx/FluentAvalonia#281] [Q&A] [answered]
**Describe the bug**
Switching themes for testing, and appears a problem with the visibility of PathIcon in FluentAvalonia. In Avalonia.Fluent theme shows correctly, but not in FluentAvalonia. Tested inside a Button an a ListBoxItem. Using DevTools the PathIcon don even show. Using ListBoxItem I can check PathIcon is there after selecting in DevTools but in the interface is transparent. Changing Background and Foreground dont have any effect. Any workaround is welcome.

**Screenshots**
![Captura de pantalla (30)](https://user-images.githubusercontent.com/27690651/215484574-fb410832-6502-429d-8cda-f41af336325f.png)
![Captura de pantalla (31)](https://user-images.githubusercontent.com/27690651/215484579-e9087dd8-5b80-4e37-bdda-e65403ff7c4b.png)

**Expected behaviour**
![Captura de pantalla (32)](https://user-images.githubusercontent.com/27690651/215486837-f5bc3a22-36e5-406f-b27f-450a005079f9.png)

**Desktop/Platform (please complete the following information):**
 - OS: Windows 10
 - FluentAvalonia Version 2.0.0-preview4
 - Avalonia Version 11-preview4

**Additional context**

**Comment by timunie** [comment, marked as answer]
FluentAvalonia has it's own `PathIcon`. I think it was renamed to `FAPathIcon` now in order to be distinct from Avalonia. If you want to render Avalonias PathIcons, you have to add a Style on your own. The original can be found here: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Themes.Fluent/Controls/PathIcon.xaml

@amwx do you want to add support for this control? If so, I can open a PR where I copy over the Style linked above.

**Comment by alex6dj** [comment]
You are right just tested before even read this comment. But I was trying to stick with the base `PathIcon` because this way I dont have to do any code change. Any way, the other thing stopping the use of this library was a performance problem I had with `ItemPresenter` some time ago and this already go away then I think is a good time to make a migration. I really like the hard work done in this. Feel free to close this issue if will stick with the internal `FAPathIcon`.

**Comment by amwx** [comment]
> @amwx do you want to add support for this control? If so, I can open a PR where I copy over the Style linked above.

Because I have an alternative, let's leave it for now, but if demand grows for it then we can do that in the future. 

@alex6dj 
I'm not sure how FA was leading to perf issues with ItemsPresenter, but ItemsControls should be getting better in preview5 after that refactor. Also, if perf is an issue, you could looking into using `IconSource` over `FAIconElement`. The idea there is similar to flyouts where you can declare the icon once and use it everywhere, which is probably really useful for your case of using them in an ItemsControl, ex:

```Xaml
<UserControl.Resources>
    <ui:PathIconSource x:Key="DeleteIcon" Data="M..." />
</UserControl.Resources>

<!-- Usage -->
<Button>
    <ui:IconSourceElement IconSource="{StaticResource DeleteIcon}" />
</Button>
```

Will also move to discussion as this may be useful to others

**Reply by alex6dj** [reply]
@amwx 
Sorry that was some month ago, maybe I was doing something wrong. Is my first app in `Avalonia`. After switching to `Avalonia.Fluent` every went fine but I really like the look of this library. So Im here again, until now everything looks ok.
About switching to `FAPathIcon`, Im ok with that, any way Im going to do a lot of changes using `FluentAvalonia`. I will test `IconSource`, thanks for the tip.

**Comment by alex6dj** [comment]
@amwx
Ok found a couple problems when using `SettingsExpander` and `PathIconSource`.
- `PathIconSource` dont work for me because destroy the icon used, at least when used inside a `SettingsExpander` not tested other way.

with `PathIconSource`
![Captura de pantalla (36)_LI](https://user-images.githubusercontent.com/27690651/215659919-2e0f7a43-f482-4b72-a787-aeb3bfd4487a.jpg)

with `FAPathIcon`
![Captura de pantalla (38)_LI](https://user-images.githubusercontent.com/27690651/215659936-8654dce9-c3d6-4320-8443-2712e40daf71.jpg)

- Putting a button inside a `SettingsExpander.Footer`, the button datacontext type is correct and the binding seems correct but the button dont execute the command. Putting the Button outside of `SettingsExpander.Footer` solve the problem.
 ```csharp
<controls:SettingsExpander.Footer>
    <Button Command="{CompiledBinding Test}">butttoooon</Button>
</controls:SettingsExpander.Footer>
```

- The last one is about the Chevron (I thinks this is the name of the arrow pointing down at the end of the SettingsExpander). When I jump to the view containing the SettingsExpander if the expander dont contains any item the chevron shows for a few ms an then dissapears. This just looks uggly.

Sorry my english is not the best, but hope you understand what I said. If any other thing is needed, code, image, video... just ask.

**Reply by amwx** [reply]
> `PathIconSource` dont work for me because destroy the icon used, at least when used inside a `SettingsExpander` not tested other way.

Try using the `Stretch`/`StretchDirection` properties. IIRC, FAPathIcon sets defaults them to Stretch.Uniform/StretchDirection.Both, but I don't think I set the defaults in PathIconSource (so its probably defaulting to no stretch behavior and drawing the path at its native side)


>     * Putting a button inside a `SettingsExpander.Footer`, the button datacontext type is correct and the binding seems correct but the button dont execute the command. Putting the Button outside of `SettingsExpander.Footer` solve the problem.
> 
> 
> ```cs
> <controls:SettingsExpander.Footer>
>    <Button Command="{CompiledBinding Test}">butttoooon</Button>
> </controls:SettingsExpander.Footer>
> ```

Hmm ok. I'll have a look at that tomorrow.

> 
>     * The last one is about the Chevron (I thinks this is the name of the arrow pointing down at the end of the SettingsExpander). When I jump to the view containing the SettingsExpander if the expander dont contains any item the chevron shows for a few ms an then dissapears. This just looks uggly.

Also interesting. Will also have a look at this, though this one may be with the styling system.

SettingsExpander is still pretty new (added in one of the previews) so it's still going through its "does it work properly" phase, but this does help!

**Reply by timunie** [reply]
> APathIcon sets defaults them to Stretch.Uniform/StretchDirection.Both, but I don't think I set the defaults in PathIconSource

Is this intended? Otherwise I think we should fix it and align it to FAPathIcon.

**Reply by amwx** [reply]
Ok, all of these issues should now be resolved in the preview5 branch. 
For the button's command not firing, it turns out raising that click event on the footer button, was triggering the Click event to be raised on the ToggleButton within SettingsExpander (from the actual Expander control in the template) where I have a handler that marks the event as handled. So I've added a check to make sure the source of the event matches the toggle button and ignores it otherwise. I think this may be a bug in the event code, but I'd have to check with WPF first.

For the expander chevron, based on your description I figured it would be challenging to repro that, so what I've done instead is just hide the chevron by default and only show it when items are added, so hopefully that will help.

**Reply by alex6dj** [reply]
Excellent. I will continue changes and waitting for the new preview release. Thanx for the hard work. I will post anything new I found

**Comment by alex6dj** [comment]
@amwx
I think I have another request more about `SettingsExpander`. Sorry for the troubles.

![Captura de pantalla (40)](https://user-images.githubusercontent.com/27690651/216068139-d52fb163-880d-45e4-84fa-d38f185d73a3.png)

Because I can bind a list of items to `SettingsExpander` then Im using it like a container. The request (I think I can do it in my side) is about the chevron (down arrow) and the animation of the control when it can expand. I think the visibility of the chevron and the animation (when you put your mouse over a `SettingsExpander` wich can expand) should react to the number of items. If I delete all the items then there is nothing to show (`0` `SettingsExpanderItem`) I think the chevron should be hidden and the animation should stop, so when you put your mouse over the control you know its empty. The same the other way, when it start empty and you add an item from view model.

The only problem I can see is with the `IsExpanded` property, because maybe the developer see it like a hint, if the `IsExpanded=true` the control should get expanded when the first item get added.

Again, thanks for your time.

**Reply by alex6dj** [reply]
Im still working with preview4, so maybe this is already done in preview5 branch. Just checking.

**Reply by alex6dj** [reply]
Another detail, the control work ok when the view is opened the first time, this is more about dynamic state changes.

**Comment by alex6dj** [comment]
Here again 😉
![Captura de pantalla (44)](https://user-images.githubusercontent.com/27690651/216095768-0e843769-2b46-4098-902a-57d6835ea79e.png)

I think this conversation title should be renamed or moved.

`SettingsExpander` again, now disabled Style I think. In the picture `SettingsExpander` and `SettingsExpanderItem` are `disabled` but only SettingsExpander show the correct behaviour in the `IconSource` and `Header/Content`. Footer is working ok.

**Comment by amwx** [comment]
Answering both here:

First: 
> I think this conversation title should be renamed or moved.

If you find bugs, open a new issue (or new discussion if its not an issue/feature request), which also allows it to be tracked better. You can combine things into one issue if you find multiple rather than opening one for each. (You don't need to do that for anything already here)

For the first issue:
That should be fixable with a pseudoclass adjustment tied to the `ItemsCollectionChanged` callback from ItemsControl. Currently you'll have to completely reset the Items property for it to take effect.

For the second issue: 
That's a bug in that I've accidentally put the disabled styles in the ":allowClick" pseudoclass selector. That's what happens when you go on autopilot making control styles 😂
